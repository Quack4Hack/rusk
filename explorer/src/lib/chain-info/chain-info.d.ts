type BlockHeader = {
  date: Date;
  feespaid: number;
  hash: string;
  height: number;
  nextblockhash: string;
  prevblockhash: string;
  reward: number;
  seed: string;
  statehash: string;
};

type BlockTransactions = {
  data: Transaction[];
  stats: {
    averageGasPrice: number;
    gasLimit: number;
    gasUsed: number;
  };
};

type Block = {
  header: BlockHeader;
  transactions: BlockTransactions;
};

type ChainInfo = {
  blocks: Block[];
  transactions: Transaction[];
};

type SearchResult = {
  id: string;
  type: "block" | "transaction";
};

type Transaction = {
  blockhash: string;
  blockheight: number;
  contract: string;
  date: Date;
  feepaid: number;
  gaslimit: number;
  gasprice: number;
  gasspent: number;
  method: string;
  success: boolean;
  txerror: string;
  txid: string;
};
