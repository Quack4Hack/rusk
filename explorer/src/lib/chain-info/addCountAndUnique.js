/**
 * Returns an array of unique objects, sorted by the frequency of their occurrence,
 * with an additional count property representing the number of times
 * that city-country combination appeared in the input data.
 *
 * @param {Omit<NodeLocation, "count">[]} data
 * @returns {NodeLocation[]}
 */
function addCountAndUnique(data) {
  const countMap = data.reduce((locationMap, location) => {
    const key = `${location.city},${location.country}`;

    if (!locationMap[key]) {
      locationMap[key] = { ...location, count: 0 };
    }
    locationMap[key].count += 1;
    return locationMap;
  }, /** @type {{ [key: string]: NodeLocation }} */ ({}));

  // Return only unique city-country combinations
  return Object.values(countMap).sort((a, b) => b.count - a.count);
}

export default addCountAndUnique;
