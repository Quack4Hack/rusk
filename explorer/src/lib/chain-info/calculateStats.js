import {
  always,
  compose,
  condition,
  filterWith,
  getKey,
  partition,
} from "lamb";

import { arraySumByKey } from "$lib/dusk/array";

/** @type {(txs: Pick<GQLTransaction, "err">[]) => number} */
const getFailedTxAmount = compose(
  getKey("length"),
  filterWith((tx) => tx.err !== null)
);

/**
 * We take into account only provisioners with
 * the minimum stake amount of 1000 Dusk (1e12 Lux).
 *
 * @type {(provisioners: HostProvisioner[]) => HostProvisioner[]}
 */
const getValidProvisioners = filterWith((p) => p.amount >= 1e12);

/**
 * Sums the values of the "amount" key in the received array.
 * Returns zero if the array is empty.
 */
const sumByAmount = condition(
  (provisioners) => provisioners.length > 0,
  arraySumByKey("amount"),
  always(0)
);

/**
 * @param {HostProvisioner[]} provisioners
 * @param {number} lastHeight
 * @param {Pick<GQLTransaction, "err">[]} last100BlocksTxs
 * @returns {Stats}
 */
function calculateStats(provisioners, lastHeight, last100BlocksTxs) {
  const [activeProvisioners, waitingProvisioners] = partition(
    getValidProvisioners(provisioners),
    (p) => p.eligibility <= lastHeight
  );

  return {
    activeProvisioners: activeProvisioners.length,
    activeStake: sumByAmount(activeProvisioners),
    lastBlock: lastHeight,
    txs100blocks: {
      failed: getFailedTxAmount(last100BlocksTxs),
      transfers: last100BlocksTxs.length,
    },
    waitingProvisioners: waitingProvisioners.length,
    waitingStake: sumByAmount(waitingProvisioners),
  };
}

export default calculateStats;
