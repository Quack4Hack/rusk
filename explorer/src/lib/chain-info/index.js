export { default as transformBlock } from "./transformBlock";
export { default as transformTransaction } from "./transformTransaction";
