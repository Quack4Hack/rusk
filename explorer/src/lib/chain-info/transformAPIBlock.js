import { mapWith, pipe, skip, updateIn } from "lamb";

import { unixTsToDate } from "$lib/dusk/date";

import { transformAPITransaction } from ".";

/** @type {(v: APIBlockHeader) => Required<APIBlockHeader>} */
const mergeWithDefaults = (v) => ({
  nextblockhash: "",
  prevblockhash: "",
  statehash: "",
  ...v,
});

/** @type {(header: Required<APIBlockHeader>) => BlockHeader} */
const addHeaderDate = (header) => ({
  ...header,
  date: unixTsToDate(header.ts),
});

/** @type {(v: APIBlockHeader) => BlockHeader} */
const transformAPIBlockHeader = pipe([
  mergeWithDefaults,
  addHeaderDate,
  skip(["__typename", "timestamp", "ts", "version"]),
]);

/** @type {(v: APIBlock) => Block} */
const transformAPIBlock = ({ header, transactions }) => ({
  header: transformAPIBlockHeader(header),
  transactions: updateIn(
    transactions,
    "data",
    mapWith(transformAPITransaction)
  ),
});

export default transformAPIBlock;
