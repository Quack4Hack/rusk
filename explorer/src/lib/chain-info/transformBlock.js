import { mapWith, pipe, skip, updateIn } from "lamb";

import { unixTsToDate } from "$lib/dusk/date";

import { transformTransaction } from ".";

/** @type {(v: APIBlockHeader) => Required<APIBlockHeader>} */
const mergeWithDefaults = (v) => ({
  nextblockhash: "",
  prevblockhash: "",
  statehash: "",
  ...v,
});

/** @type {(header: Required<APIBlockHeader>) => BlockHeader} */
const addHeaderDate = (header) => ({
  ...header,
  date: unixTsToDate(header.ts),
});

/** @type {(v: APIBlockHeader) => BlockHeader} */
const transformBlockHeader = pipe([
  mergeWithDefaults,
  addHeaderDate,
  skip(["__typename", "timestamp", "ts"]),
]);

/** @type {(v: APIBlock) => Block} */
const transformBlock = ({ header, transactions }) => ({
  header: transformBlockHeader(header),
  transactions: updateIn(transactions, "data", mapWith(transformTransaction)),
});

export default transformBlock;
