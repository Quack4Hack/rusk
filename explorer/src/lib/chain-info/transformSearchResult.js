import { compose, getPathIn, mapWith } from "lamb";

/**
 * @template {Record<PropertyKey, any>} T
 * @template {string} U
 * @param {T} source
 * @param {U} path
 * @returns {Exclude<import("lamb").GetPath<T, U>, undefined> | []}
 */
const getPathOrEmptyArray = (source, path) => getPathIn(source, path) ?? [];

/** @type {(v: APISearchResult) => SearchResult[]} */
const transformSearchResult = compose(
  mapWith((entry) =>
    entry?.header?.hash
      ? { id: entry.header.hash, type: "block" }
      : { id: entry.txid, type: "transaction" }
  ),
  (v) => [
    ...getPathOrEmptyArray(v, "data.data.blocks"),
    ...getPathOrEmptyArray(v, "data.data.transactions"),
  ]
);

export default transformSearchResult;
