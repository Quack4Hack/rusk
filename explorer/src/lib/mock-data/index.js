export { default as apiMarketData } from "./api-market-data.json";
export { default as apiNodeLocations } from "./api-node-locations.json";
export { default as apiStats } from "./api-stats.json";
export { default as gqlBlock } from "./gql-block.json";
export { default as gqlBlocks } from "./gql-blocks.json";
export { default as gqlLatestChainInfo } from "./gql-latest-chain-info.json";
export { default as gqlSearchPossibleResults } from "./gql-search-possible-results.json";
export { default as gqlTransaction } from "./gql-transaction.json";
export { default as gqlTransactionDetails } from "./gql-transaction-details.json";
export { default as gqlTransactions } from "./gql-transactions.json";
