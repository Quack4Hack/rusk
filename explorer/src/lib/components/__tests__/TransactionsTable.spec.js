import { afterAll, afterEach, describe, expect, it, vi } from "vitest";
import { cleanup, render } from "@testing-library/svelte";
import { mapWith, slice } from "lamb";

import { gqlTransactions } from "$lib/mock-data";
import { transformTransaction } from "$lib/chain-info";

import { TransactionsTable } from "..";

describe("Transactions Table", () => {
  vi.useFakeTimers();
  vi.setSystemTime(new Date(2024, 4, 30));

  const transformTransactions = mapWith(transformTransaction);
  const data = slice(
    transformTransactions(gqlTransactions.transactions),
    0,
    10
  );

  /** @type {import("svelte").ComponentProps<TransactionsTable>} */
  const baseProps = {
    data: data,
    mode: "full",
  };
  const baseOptions = {
    props: baseProps,
    target: document.body,
  };

  afterEach(cleanup);

  afterAll(() => {
    vi.useRealTimers();
  });

  it('should render the `TransactionsTable` component in "full" mode', () => {
    const { container } = render(TransactionsTable, baseOptions);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the `TransactionsTable` component in "compact" mode', () => {
    /** @type {import("svelte").ComponentProps<TransactionsTable>} */
    const props = { ...baseProps, mode: "compact" };
    const { container } = render(TransactionsTable, { ...baseOptions, props });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should pass additional class names to the rendered element", () => {
    const props = { ...baseProps, className: "foo bar" };
    const { container } = render(TransactionsTable, { ...baseOptions, props });

    expect(container.firstChild).toMatchSnapshot();
  });
});
