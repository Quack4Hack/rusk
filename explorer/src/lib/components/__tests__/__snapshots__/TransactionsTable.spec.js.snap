// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Transactions Table > should render the \`TransactionsTable\` component 1`] = `
<div
  class="table-container transactions-table"
>
  <table
    class="table"
  >
    <thead
      class="table__head"
    >
      <tr
        class="table__row"
      >
        <th
          class="table__header-cell"
        >
          Hash
        </th>
        
        <!--&lt;TableCell&gt;-->
         
        <th
          class="table__header-cell"
        >
          Gas
        </th>
        
        <!--&lt;TableCell&gt;-->
         
        <th
          class="table__header-cell"
        >
          Fee
        </th>
        
        <!--&lt;TableCell&gt;-->
         
        <th
          class="table__header-cell"
        >
          Status
        </th>
        
        <!--&lt;TableCell&gt;-->
         
        <th
          class="table__header-cell"
        >
          Type
        </th>
        
        <!--&lt;TableCell&gt;-->
      </tr>
      <!--&lt;TableRow&gt;-->
    </thead>
    <!--&lt;TableHead&gt;-->
     
    <tbody
      class="table__body"
    >
      
      <tr
        class="table__row"
      >
        <td
          class="table__data-cell"
        >
          <a
            class="dusk-anchor transaction__link"
            href="/some-base-path/transactions/transaction?id=4877687c2dbf154248d3ddee9ba0d81e3431f39056f82a46819da041d4ac0e04"
          >
            3c6e4018cf...308e07bd2d
          </a>
          <!--&lt;Anchor&gt;-->
          <!--&lt;AppAnchor&gt;-->
           
          <small
            class="transaction__time"
          >
            last month
          </small>
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <b
            class="transaction__fee-price-label"
          >
            PRICE:
          </b>
           
          1
          <br />
           
          <b
            class="transaction__fee-limit-label"
          >
            LIMIT:
          </b>
           
          500,000,000
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-alt"
          >
            0.000290766 Dusk
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-success"
          >
            success
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-neutral"
          >
            transfer
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
      </tr>
      <!--&lt;TableRow&gt;-->
      
      <tr
        class="table__row"
      >
        <td
          class="table__data-cell"
        >
          <a
            class="dusk-anchor transaction__link"
            href="/some-base-path/transactions/transaction?id=807d4f596ea083215bb38decf8b7e824e3efae734d6a43c2549717e57d04cf02"
          >
            40d3e59a9b...5e8ade4d3b
          </a>
          <!--&lt;Anchor&gt;-->
          <!--&lt;AppAnchor&gt;-->
           
          <small
            class="transaction__time"
          >
            last month
          </small>
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <b
            class="transaction__fee-price-label"
          >
            PRICE:
          </b>
           
          1
          <br />
           
          <b
            class="transaction__fee-limit-label"
          >
            LIMIT:
          </b>
           
          500,000,000
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-alt"
          >
            0.000290053 Dusk
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-success"
          >
            success
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-neutral"
          >
            transfer
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
      </tr>
      <!--&lt;TableRow&gt;-->
      
      <tr
        class="table__row"
      >
        <td
          class="table__data-cell"
        >
          <a
            class="dusk-anchor transaction__link"
            href="/some-base-path/transactions/transaction?id=744b1d303f71e76e530b0543c284f65a30a26b75eab194d473c59d7fbd676a0b"
          >
            ea86c56ffa...8914074fe3
          </a>
          <!--&lt;Anchor&gt;-->
          <!--&lt;AppAnchor&gt;-->
           
          <small
            class="transaction__time"
          >
            last month
          </small>
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <b
            class="transaction__fee-price-label"
          >
            PRICE:
          </b>
           
          1
          <br />
           
          <b
            class="transaction__fee-limit-label"
          >
            LIMIT:
          </b>
           
          500,000,000
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-alt"
          >
            0.000290219 Dusk
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-success"
          >
            success
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-neutral"
          >
            transfer
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
      </tr>
      <!--&lt;TableRow&gt;-->
      
      <tr
        class="table__row"
      >
        <td
          class="table__data-cell"
        >
          <a
            class="dusk-anchor transaction__link"
            href="/some-base-path/transactions/transaction?id=34b4404adc0d286e62ddb9bf80429d4491b61de3f5e4c54975724a6d248d8e0b"
          >
            ea86c56ffa...8914074fe3
          </a>
          <!--&lt;Anchor&gt;-->
          <!--&lt;AppAnchor&gt;-->
           
          <small
            class="transaction__time"
          >
            last month
          </small>
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <b
            class="transaction__fee-price-label"
          >
            PRICE:
          </b>
           
          1
          <br />
           
          <b
            class="transaction__fee-limit-label"
          >
            LIMIT:
          </b>
           
          500,000,000
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-alt"
          >
            0.00029046 Dusk
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-success"
          >
            success
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-neutral"
          >
            transfer
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
      </tr>
      <!--&lt;TableRow&gt;-->
      
      <tr
        class="table__row"
      >
        <td
          class="table__data-cell"
        >
          <a
            class="dusk-anchor transaction__link"
            href="/some-base-path/transactions/transaction?id=88ac1818eb5aeee47a1356fe5d75726943c12ea1fb8ed23efb37c847a318ed0e"
          >
            a85a0f5cc2...f0bb827312
          </a>
          <!--&lt;Anchor&gt;-->
          <!--&lt;AppAnchor&gt;-->
           
          <small
            class="transaction__time"
          >
            last month
          </small>
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <b
            class="transaction__fee-price-label"
          >
            PRICE:
          </b>
           
          1
          <br />
           
          <b
            class="transaction__fee-limit-label"
          >
            LIMIT:
          </b>
           
          500,000,000
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-alt"
          >
            0.000291263 Dusk
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-success"
          >
            success
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-neutral"
          >
            transfer
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
      </tr>
      <!--&lt;TableRow&gt;-->
      
      <tr
        class="table__row"
      >
        <td
          class="table__data-cell"
        >
          <a
            class="dusk-anchor transaction__link"
            href="/some-base-path/transactions/transaction?id=3e430f7bbc123585a21b0689331d77a8c9e3da9b75ccc463aaf22d8a581e5507"
          >
            646bd535ae...67c1cb0906
          </a>
          <!--&lt;Anchor&gt;-->
          <!--&lt;AppAnchor&gt;-->
           
          <small
            class="transaction__time"
          >
            last month
          </small>
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <b
            class="transaction__fee-price-label"
          >
            PRICE:
          </b>
           
          1
          <br />
           
          <b
            class="transaction__fee-limit-label"
          >
            LIMIT:
          </b>
           
          500,000,000
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-alt"
          >
            0.000290735 Dusk
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-success"
          >
            success
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-neutral"
          >
            transfer
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
      </tr>
      <!--&lt;TableRow&gt;-->
      
      <tr
        class="table__row"
      >
        <td
          class="table__data-cell"
        >
          <a
            class="dusk-anchor transaction__link"
            href="/some-base-path/transactions/transaction?id=776c0c54ccdd485fb32d6562b2b8b10f3cb283465da6eecba861f384725e3f0b"
          >
            fbcda951c6...4bec844f78
          </a>
          <!--&lt;Anchor&gt;-->
          <!--&lt;AppAnchor&gt;-->
           
          <small
            class="transaction__time"
          >
            last month
          </small>
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <b
            class="transaction__fee-price-label"
          >
            PRICE:
          </b>
           
          1
          <br />
           
          <b
            class="transaction__fee-limit-label"
          >
            LIMIT:
          </b>
           
          500,000,000
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-alt"
          >
            0.000289627 Dusk
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-success"
          >
            success
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-neutral"
          >
            transfer
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
      </tr>
      <!--&lt;TableRow&gt;-->
      
      <tr
        class="table__row"
      >
        <td
          class="table__data-cell"
        >
          <a
            class="dusk-anchor transaction__link"
            href="/some-base-path/transactions/transaction?id=e022261204125d87ad99d404b73d0ca3ce4c0985e50f08b15a1619248ae0f702"
          >
            a90821aca4...11467eb9c6
          </a>
          <!--&lt;Anchor&gt;-->
          <!--&lt;AppAnchor&gt;-->
           
          <small
            class="transaction__time"
          >
            last month
          </small>
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <b
            class="transaction__fee-price-label"
          >
            PRICE:
          </b>
           
          1
          <br />
           
          <b
            class="transaction__fee-limit-label"
          >
            LIMIT:
          </b>
           
          500,000,000
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-alt"
          >
            0.00029061 Dusk
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-success"
          >
            success
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-neutral"
          >
            transfer
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
      </tr>
      <!--&lt;TableRow&gt;-->
      
      <tr
        class="table__row"
      >
        <td
          class="table__data-cell"
        >
          <a
            class="dusk-anchor transaction__link"
            href="/some-base-path/transactions/transaction?id=fd9af8e161863389c2c52c900dc3881c754c7d2549f865df95ccfa0f57fe5b04"
          >
            4db9e793b1...e2ab163e74
          </a>
          <!--&lt;Anchor&gt;-->
          <!--&lt;AppAnchor&gt;-->
           
          <small
            class="transaction__time"
          >
            last month
          </small>
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <b
            class="transaction__fee-price-label"
          >
            PRICE:
          </b>
           
          1
          <br />
           
          <b
            class="transaction__fee-limit-label"
          >
            LIMIT:
          </b>
           
          500,000,000
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-alt"
          >
            0.000291347 Dusk
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-success"
          >
            success
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-neutral"
          >
            transfer
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
      </tr>
      <!--&lt;TableRow&gt;-->
      
      <tr
        class="table__row"
      >
        <td
          class="table__data-cell"
        >
          <a
            class="dusk-anchor transaction__link"
            href="/some-base-path/transactions/transaction?id=65be1885cbbf9545bc3b0f3c56b34da79672f2be063d81a5d4f78b2ca5c18101"
          >
            47c9f1349f...e7ce5103db
          </a>
          <!--&lt;Anchor&gt;-->
          <!--&lt;AppAnchor&gt;-->
           
          <small
            class="transaction__time"
          >
            last month
          </small>
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <b
            class="transaction__fee-price-label"
          >
            PRICE:
          </b>
           
          1
          <br />
           
          <b
            class="transaction__fee-limit-label"
          >
            LIMIT:
          </b>
           
          500,000,000
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-alt"
          >
            0.000290416 Dusk
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-success"
          >
            success
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
        <td
          class="table__data-cell"
        >
          <span
            class="dusk-badge dusk-badge--variant-neutral"
          >
            transfer
          </span>
          <!--&lt;Badge&gt;-->
        </td>
        
        <!--&lt;TableCell&gt;-->
         
      </tr>
      <!--&lt;TableRow&gt;-->
      
    </tbody>
    <!--&lt;TableBody&gt;-->
  </table>
</div>
`;
